### PR TITLE
Implement common::memory::Memory

### PR DESCRIPTION
**Pull request by stratact**
@jackpot51 does everything look good?

### CODE REVIEW

**Comment by jackpot51** [comment]
It needs unsafe wrapping

File: src/common/memory.rs [hunk 1]
```diff
@@ -21,6 +21,34 @@ struct MemoryMapEntry {
...
+    pub fn new(size: usize) -> Option<Self> {
+        let alloc = alloc(size);
+        if alloc > 0 {
+            Some(Memory { address: alloc })
+        }
+        else { None }
+    }
+
+    pub fn renew(&self, size: usize) -> Option<Self> {
+        let realloc = realloc(self.address, size);
```
Comment (jackpot51) [inline src/common/memory.rs]: should be let realloc = unsafe { realloc(...) };

File: src/common/memory.rs [hunk 2]
```diff
@@ -21,6 +21,34 @@ struct MemoryMapEntry {
     acpi: u32,
 }
 
+struct Memory {
+    address: usize,
+}
+
+impl Memory {
+    pub fn new(size: usize) -> Option<Self> {
+        let alloc = alloc(size);
```
Comment (jackpot51) [inline src/common/memory.rs]: let alloc = unsafe { alloc(size) };

File: src/common/memory.rs [hunk 3]
```diff
@@ -21,6 +21,34 @@ struct MemoryMapEntry {
...
+        if realloc > 0 {
+            Some(Memory { address: realloc })
+        }
+        else { None }
+    }
+}
+
+impl Drop for Memory {
+    fn drop(&mut self) {
+        unalloc(self.address)
```
Comment (jackpot51) [inline src/common/memory.rs]: unsafe{ unalloc(self.address) }

File: src/common/memory.rs (hunk 1 above)
Comment (stratact) [inline src/common/memory.rs]: actually, I tried that, and rustc didn't like that

Comment (stratact) [inline src/common/memory.rs]: oh wait, nevermind. I'm stupid. What you suggest did work after trying it again.

Comment (jackpot51) [inline src/common/memory.rs]: Either way looks fine, now its time to propogate. I'm gonna take src/common/vec.rs

**Pull request merged**